torch/utils/flop_counter.py

```diff
@@ -91,58 +83,96 @@ def conv_flop_count(
-
-    return conv_flop_count(x_shape, w_shape, out_shape, transposed=transposed)
+    return conv_flop_count(x_shape, w_shape, out, transposed=transposed)
 
 def transpose_shape(shape):
     return [shape[1], shape[0]] + list(shape[2:])
 
-def conv_backward_flop(input_shapes: List[Any], output_shapes: List[Any]) -> int:
-    grad_out_shape, x_shape, w_shape = input_shapes[:3]
-    output_mask = input_shapes[-1]
-    fwd_transposed = input_shapes[7]
+def conv_backward_flop(
+        grad_out_shape,
+        x_shape,
+        w_shape,
+        _bias,
+        _stride,
+        _padding,
+        _dilation,
+        transposed,
+        _output_padding,
+        _groups,
+        output_mask,
+        out) -> int:
     flop_count = 0
 
     if output_mask[0]:
-        grad_input_shape = get_shape(output_shapes[0])
-        flop_count += conv_flop_count(grad_out_shape, w_shape, grad_input_shape, not fwd_transposed)
+        grad_input_shape = get_shape(out[0])
+        flop_count += conv_flop_count(grad_out_shape, w_shape, grad_input_shape, not transposed)
     if output_mask[1]:
-        grad_weight_shape = get_shape(output_shapes[1])
-        flop_count += conv_flop_count(transpose_shape(x_shape), grad_out_shape, grad_weight_shape, fwd_transposed)
+        grad_weight_shape = get_shape(out[1])
+        flop_count += conv_flop_count(transpose_shape(x_shape), grad_out_shape, grad_weight_shape, transposed)
 
     return flop_count
 
 def sdpa_flop_count(query_shape, key_shape):
     """
     Count flops for self-attention.
+    NB: We can assume that value_shape == key_shape
     """
     b, h, s, d = query_shape
-    b2, h2, s2, d2 = key_shape
-    assert b == b2 and h == h2 and d == d2
+    _b2, _h2, s2, _d2 = key_shape
+    assert b == _b2 and h == _h2 and d == _d2
     total_flops = 0
-    # [b, h, s, d] @ [b, h, d, s2] -> [b, h, s, s2]
-    total_flops += bmm_flop([(b * h, s, d), (b * h, d, s2)])
-    # [b, h, s, s2] @ [b, h, s2, d] -> [b, h, s2, d]
-    total_flops += bmm_flop([(b * h, s, s2), (b * h, s2, d)])
+    # q: [b, h, s, d] @ k: [b, h, d, s2] -> scores: [b, h, s, s2]
+    total_flops += bmm_flop((b * h, s, d), (b * h, d, s2))
+    # scores: [b, h, s, s2] @ v: [b, h, s2, d] -> out: [b, h, s, d]
+    total_flops += bmm_flop((b * h, s, s2), (b * h, s2, d))
     return total_flops
 
 
 
-def sdpa_flop(input_shapes: List[Any], _=None) -> int:
+def sdpa_flop(query_shape, key_shape, value_shape, *args, out=None, **kwargs) -> int:
     """
     Count flops for self-attention.
     """
     # NB: We aren't accounting for causal attention here
-    query_shape, key_shape, value_shape, *_ = input_shapes
     return sdpa_flop_count(query_shape, key_shape)
 
+
+def sdpa_backward_flop_count(grad_out_shape, query_shape, key_shape):
+    total_flops = 0
+    b, h, s, d = query_shape
+    _b2, _h2, s2, _d2 = key_shape
+    _b3, _h3, _s3, _d3 = grad_out_shape
+    assert b == _b2 == _b3 and h == _h2 == _h3 and d == _d2 == _d3
```

Comment:
V can have a different embed dim from q,k and mem efficient supports that

Reply:
Oh ok.